Added referencesets/search client endpoint

dcolligan opened the pull request:
Issue #76 -- obviously very WIP.  Going to be adding client endpoints to this branch as I go along...

Tried to follow the conventions established in the code already.  @jeromekelleher

jeromekelleher commented on auth.cfg:
```diff
@@ -0,0 +1,2 @@
+[google]
+key = YOUR_KEY_HERE
```
I'm not crazy about adding in conf files into our source tree. What do we do when we install using `pip`? Do we try to distribute these files into standard locations? 

I don't think we really need a conf file here in any case: a CLI argument would seem the simpler choice.

dcolligan replied:
I agree that pip installs make including cfg files a bad idea.

On the other hand, we have run into this problem of configuration before, some of which can get quite complicated (logging conf, server conf, auth conf -- which could be arbitrarily complex for certain sites) and I don't think "add another command line flag" is the solution to all of them.  To me, it makes more sense to have _some_ file(s) somewhere that, yes, a user can point to with a command line argument.

The "pre-populate configs in some .py file and have a cmd line flag to select from amongst them" solution doesn't work in cases where, like auth, we do not know the values in advance.

I'll add a command line flag for this for the time being, but we really need a better solution, IMHO.

kerrydc replied:
I think we should go with flags, with sensible defaults, and the ability to
set flags via a flag specifying a config file. Note that I may be biased by
regularly setting 50-100 command line flags in a build. Longer term we will
want a guided setup that will prompt the user for various values on
installation.
Kerry

On Fri, Jan 23, 2015 at 10:29 AM, Danny Colligan <notifications@github.com>
wrote:

> In auth.cfg <https://github.com/ga4gh/server/pull/90#discussion_r23469102>
> :
>
> > @@ -0,0 +1,2 @@
> > +[google]
> > +key = YOUR_KEY_HERE
>
> I agree that pip installs make including cfg files a bad idea.
>
> On the other hand, we have run into this problem of configuration before,
> some of which can get quite complicated (logging conf, server conf, auth
> conf -- which could be arbitrarily complex for certain sites) and I don't
> think "add another command line flag" is the solution to all of them. To
> me, it makes more sense to have *some* file(s) somewhere that, yes, a
> user can point to with a command line argument.
>
> The "pre-populate configs in some .py file and have a cmd line flag to
> select from amongst them" solution doesn't work in cases where, like auth,
> we do not know the values in advance.
>
> I'll add a command line flag for this for the time being, but we really
> need a better solution, IMHO.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ga4gh/server/pull/90/files#r23469102>.
>

jeromekelleher commented on ga4gh/cli.py:
```diff
@@ -127,6 +127,31 @@ def printVariant(self, v):
+    """
+    def __init__(self, args):
+        rsr = protocol.GASearchReferenceSetsRequest()
+        if rsr.accessions:
```
I think we should prefer `if x is not None` over `if x`, as it's more explicit.

jeromekelleher commented on ga4gh/cli.py:
```diff
@@ -127,6 +127,31 @@ def printVariant(self, v):
+            rsr.accessions = args.accessions.split(",")
+        # TODO google-specific assemblyId not a valid field,
+        # referenceSetId (contrary to goog docs) not implemented
+        del rsr.__dict__['assemblyId']
```
It looks like we'll need to put in some workarounds for Google in general here. However, we don't want to make our client Google specific. Perhaps we could add in a command line argument to turn on Google workarounds?

jeromekelleher commented on ga4gh/cli.py:
```diff
@@ -127,6 +127,31 @@ def printVariant(self, v):
+        # referenceSetId (contrary to goog docs) not implemented
+        del rsr.__dict__['assemblyId']
+        # rsr.referenceSetId = args.assembly_id
+        if rsr.md5checksums:
```
Why are we doing the split here?

jeromekelleher commented on ga4gh/cli.py:
```diff
@@ -127,6 +127,31 @@ def printVariant(self, v):
+    Runner class for the referencesets/search method.
+    """
+    def __init__(self, args):
+        rsr = protocol.GASearchReferenceSetsRequest()
```
`request` is probably a better name here.

jeromekelleher commented:
Looks good @dcolligan, thanks. I've made a few comments in line, but they're pretty minor.

We need some test cases for this client code now, and we might as well start putting them in now. Have you any ideas for how we might unit test this code (i.e., the `client.searchX` functions)?

Let's take the compliance tester bit out, and keep that for a separate PR.

dcolligan commented:
Pushed a change that addresses some of your concerns.

As far as testing, the search* functions don't do much; most of the logic is in `runRequest`.  For testing it, we should probably just mock the request library calls.  Ah, and I see I've made the test coverage fail.

Although, as far as APIs go, I find end-to-end tests generally the most useful since the major problem is testing that all of the endpoints/protocol implementations line up -- there's not a ton of logic to be checked.

dcolligan commented:
I pushed some unit tests, @jeromekelleher, so let me know what you think.

jeromekelleher commented on ga4gh/client.py:
```diff
@@ -11,15 +11,19 @@
 
 import ga4gh.protocol as protocol
 
+requests.packages.urllib3.disable_warnings()
```
Can we get a comment to explain why we're disabling warnings here please?

jeromekelleher commented on tests/test_client.py:
```diff
@@ -0,0 +1,117 @@
+from __future__ import unicode_literals
+
+import unittest
+import mock
```
The convention I've adopted so far with imports is to have a number of sections. (1) futures; (2) standard libary; (3) third party libraries; (4) package local. Since mock isn't part of that standard library until Python 3.3, which we're not supporting yet, so we should put this into third party stuff. Also, I think we'll need to update the Travis script to install mock for testing.

dcolligan replied:
The CI build wouldn't have passed if we needed to explicitly `pip install mock`

jeromekelleher commented on tests/test_client.py:
```diff
@@ -0,0 +1,117 @@
+                request, url, protocolClass, listAttr)]
+
+            # verify results of invocation
+            assert len(result) == 2
```
These hard-coded values seem a bit brittle and mysterious when reading them. Can we make some constants and use these instead perhaps?

pashields replied:
In this case, they are referring to values inside the dummy response. Testing against specific values is more brittle, but it saves you from figuring out that two incorrect values are equal.

In production code I'd be with you 100%, but in tests I think this way is way safer.

jeromekelleher commented on ga4gh/cli.py:
```diff
@@ -127,6 +143,32 @@ def printVariant(self, v):
+        request = protocol.GASearchReferenceSetsRequest()
+        if args.accessions is not None:
+            request.accessions = args.accessions.split(",")
+        if self.usingWorkaroundsFor('google'):
```
I'm not crazy about using hard-coded values for program flow, as it's very brittle. Perhaps we could add a constant to `client.py` that we could check against here and in `client.py` instead?

jeromekelleher commented:
This looks great @dcolligan, thanks. I especially like the testing setup using `mock`, this seems very neat. I'm struggling to get my head around exactly how it works though - do you think you could add a few quick comments explaining the overall testing strategy and how it works? 

I'm like the workarounds API too, minus the hard-coded strings. We'll probably need something a bit more sophisticated later, but this'll get us up and running.

dcolligan commented:
Pushed another commit which squashes the previous ones and addresses some concerns here.

@jeromekelleher as far as the tests go, mocks are essentially used to provide a test stand-in for functionality that may be either expensive or impossible to set up.  Classic examples include database connections, network calls and large file calls.  Here in `TestRunRequest` I'm patching the `requests.post` call to be intercepted by the mock library and return a `DummyResponse` object.  In this way, we both preserve the property that the test suite can be run locally (thereby improving modularity and robustness since relying on a remote resource which may not be available would be a bad idea anyways) and we save some time by not actually waiting for the network.  We also improve modularity by abstracting the details of the actual requests.post library call -- some bug in a change in the requests library implementation now won't affect our test code.

In `TestSearchMethodsCallRunRequest` the idea is similar, except we don't really care to test the return value since there's no logic between us providing a dummy one and us testing what the return value would be.  So we simply assert that the method we wanted to test called the function it should be calling correctly.  Again, we're using mocks to test only the layer of code we want to test and not anything below it.  

Let me know if you still have questions about any of this.

(the only test that failed in CI is the CORS test)

jeromekelleher commented on tests/test_client.py:
```diff
@@ -0,0 +1,118 @@
+from __future__ import unicode_literals
+
+import unittest
+import requests
```
requests is a third party lib - could you move it into the next section please?

jeromekelleher commented on tests/test_client.py:
```diff
@@ -0,0 +1,118 @@
+            mockPost.side_effect = [DummyResponse(), DummyResponse('{}')]
+            request = DummyRequest()
+
+            # invoke SUT
```
What's SUT? Sorry if I'm being a bit dense here!

pashields replied:
System Under Test

jeromekelleher commented:
Thanks for the explanation @dcolligan, this helps a lot. This is a great approach, and gives a nice lightweight way of testing just what we want to test. Very nice!

I'm +1 to merge, modulo a couple of tiny nits above.

dcolligan commented:
Ok @jeromekelleher pushed something to address all your concerns (except the one that you and @pashields disagree on... I'm ambivalent about that issue myself)

pashields commented:
@jeromekelleher you have a strong feeling on the testing issue? I don't feel the need to block this PR on it either way.

jeromekelleher commented:
Sorry, should has responded to this. I'm happy to go with your view @pashields, I don't feel strongly about this at all.

pashields commented:
Cool. Thanks.

+1 on this PR

jeromekelleher commented:
Excellent. I kicked off a fresh travis build as there has been a fair bit of merging today. I'll merge once the build completes - looks like it's failing when we try to download pysam. We really need to get some binary wheels for travis sorted out!

dcolligan commented:
Yeah, I had to force push again to get Travis working... but it passes now.

jeromekelleher commented:
Merfged - thanks @dcolligan!
